docs/proposals/failover/optimize-cluster-failover.md

```diff
@@ -0,0 +1,635 @@
+
+5. Failover FeatureGate
+
+The Failover FeatureGate can impose overall constraints on the cluster failure migration mechanism. Once the Failover FeatureGate is turned off, the cluster failure judgment will no longer take effect, and no resources will be migrated.
+
+In light of the above design principles, it has been sorted out that the following parts in the current system do not meet the objectives of this proposal:
+1. In the judgment of cluster failure conditions, the "taint cluster by conditions" rule is not provided for users to define.
+2. The design of several types of taints is incomplete and needs to be supplemented.
+3. There is no application-level cluster failure migration strategy provided.
+4. The current eviction queue lacks the ability to be downgraded, and the queue does not provide metrics for administrators.
+5. The Failover FeatureGate cannot impose overall constraints on the cluster failure mechanism. Users can still trigger resource migration by adding taints.
+
+Next, let's introduce the design solutions for each of the above parts that do not meet the objectives one by one.
+
+### Cluster failure condition judgment
+
+Cluster failure condition judgment involves assessing the status conditions of a faulty cluster object to add various taints to the cluster. This process is referred to as "taint cluster by conditions".
+
+This proposal adds a new API: ClusterTaintPolicy, which can be used to describe what taints should be added to the target cluster when certain cluster status conditions are met. The relevant logic can be implemented by adding a new controller. This controller is responsible for monitoring changes in the cluster status conditions, that is, the `.status.conditions` field of the Cluster object, and then deciding how to modify the taints of the Cluster object according to the ClusterTaintPolicy API defined by the user. When the matching conditions are no longer met, the taints added to the target cluster will be removed.
+
+Key design considerations for ClusterTaintPolicy:
+- To avoid over-design, ClusterTaintPolicy currently mainly provides condition judgment based on `cluster.status.conditions` as input. More input conditions can be added as needed in the future.
+- When making input condition judgments based on `cluster.status.conditions`, in addition to `condition.type` and `condition.status`, time-dimension matching is also required to prevent taint jitter caused by frequent changes in cluster conditions and make the system overly sensitive. However, also to avoid over-design, in the initial version, only simple time-window-based judgments can be provided. For example, if the condition indicating that the cluster is disconnected persists for 5 minutes, add a taint; if the condition indicating that the cluster has re - established connection holds for 3 minutes, remove the taint.
```

Comment:
```suggestion
- When making input condition judgments based on `cluster.status.conditions`, in addition to `condition.type` and `condition.status`, time-dimension matching is also required to prevent taint jitter caused by frequent changes in cluster conditions and make the system overly sensitive. However, also to avoid over-design, in the initial version, only simple time-window-based judgments will be provided. For example, if the condition indicating that the cluster is disconnected persists for 5 minutes, add a taint; if the condition indicating that the cluster has re-established connection holds for 3 minutes, remove the taint.
```